[Merged by Bors] - feat(linear_algebra/basic): curry is linear_equiv

pechersky opened the pull request:
Currying provides a linear_equiv. This can be used to show that
matrix lookup is a linear operation. This should allow to easily
provide a finite_dimensional instance for matrices.

---
<!-- put comments you want to keep out of the PR commit here -->

pechersky committed:
feat(linear_algebra/basic): curry is linear_equiv

Currying provides a linear_equiv. This can be used to show that
matrix lookup is a linear operation. This should allow to easily
provide a finite_dimensional instance for matrices.j

urkud commented on src/linear_algebra/basic.lean:
```diff
@@ -1641,6 +1641,19 @@ rfl
+
+variables (V V₂ R)
+
+/-- Linear equivalence between an uncurried and curried function. -/
```
Please add a reference to [tensor_product.curry](https://leanprover-community.github.io/mathlib_docs/linear_algebra/tensor_product.html#tensor_product.curry) in the docstring. This is a different operation but a person looking for it may find your definition first.

Usually we also add lemmas `coe_curry` and `coe_curry_symm`.

kim-em commented on src/linear_algebra/basic.lean:
```diff
@@ -1641,6 +1641,19 @@ rfl
+/-- Linear equivalence between an uncurried and curried function. -/
+protected def curry :
+  (V → V₂ → R) ≃ₗ[R] (V × V₂ → R) :=
+{ add := λ _ _, by ext z; cases z; refl,
```
The preferred style is now not to use `;` except where it is necessary --- that is, when you want to run one tactic on multiple produced goals.

Instead use `by { ext z, cases z, refl, }`, with braces.

(Lots of older code in mathlib was written before this style rule was introduced, and never updated.)

pechersky committed:
chore(linear_algebra/basic): linting semicolons

pechersky committed:
chore(linear_algebra/basic): add ref to tensor_product.curry in docstring

pechersky committed:
chore(linear_algebra/basic): rename to uncurry

Linear equivalence of curried and uncurried functions goes in the
uncurry direction based on `equiv.arrow_arrow_equiv_prod_arrow`.
This direction matches the type of `function.uncurry`, so the name
should match too.

pechersky committed:
feat(linear_algebra/basic): add coe lemmas to uncurry

pechersky committed:
chore(linear_algebra/basic): rename uncurry section

urkud commented on src/linear_algebra/basic.lean:
```diff
@@ -1641,6 +1641,24 @@ rfl
+
+@[simp] lemma coe_uncurry (x) : linear_equiv.uncurry R V V₂ x = uncurry x := rfl
+
+@[simp] lemma coe_curry_symm (x) : (linear_equiv.uncurry R V V₂).symm x = curry x := rfl
```
```suggestion
@[simp] lemma coe_uncurry : ⇑(linear_equiv.uncurry R V V₂) = uncurry := rfl

@[simp] lemma coe_uncurry_symm : ⇑(linear_equiv.uncurry R V V₂).symm = curry := rfl
```

jcommelin committed:
Update src/linear_algebra/basic.lean

Co-authored-by: Yury G. Kudryashov <urkud@urkud.name>

jcommelin reviewed (approved):
Thanks :tada:

bors merge
